### PR TITLE
Update WlSeat to version 7

### PR DESCRIPTION
**Pull request by PolyMeilex**
Ver 6 was about touch events that we already have: https://github.com/wayland-project/wayland/commit/6a18a87727c64719c68168568b9ab1e4d7c2d9c1
Ver 7 is about sealing the keymap files:  https://github.com/wayland-project/wayland/commit/905c0a341ddf0a885811d19e2b79c65a3f1d210c

The Sealing could allow for some optimizations, but I'm not sure which OSes support that, so I just left the inefficient version where we create a file per client.

### CODE REVIEW

**Comment by HeavyRain266** [comment]
As far as I'm aware, from systems where you can actually run wayland (linux & freebsd), seems like only linux support that kind of feature but on the other hand freebsd uses libinput layer and will have to look into kernel code to see if there is sealing.

File: src/wayland/seat/keyboard/mod.rs [hunk 1]
```diff
@@ -544,6 +545,13 @@ impl KeyboardHandle {
         let ret = tempfile().and_then(|mut f| {
             f.write_all(self.arc.keymap.as_bytes())?;
             f.flush()?;
+
+            if kbd.version() >= 7 {
+                if let Err(err) = nix::fcntl::fcntl(f.as_raw_fd(), FcntlArg::F_GET_SEALS) {
```
Comment (vberger) [inline src/wayland/seat/keyboard/mod.rs:550]: Is this call actually setting any kind of seals on the file?

Comment (PolyMeilex) [reply src/wayland/seat/keyboard/mod.rs:550]: My bad, the error log is misleading. This is basically a check if selling is supported, so the log should just be "Sealing not supported".

(If selling is supported, we should then return an existing fd of a sealed file)

I guess this log does not make much sense if the whole thing is not implemented.
I guess I'll implement the whole thing, it is supported on Linux and FreeBSD and that's the only targets we care about anyway.

Comment (PolyMeilex) [reply src/wayland/seat/keyboard/mod.rs]: Done, replaced with seal support

File: src/wayland/seat/keyboard/keymap_file.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        file.write_all(keymap)?;
+        file.flush()?;
+
+        file.seek(std::io::SeekFrom::Start(0))?;
+
+        nix::fcntl::fcntl(
+            file.as_raw_fd(),
+            FcntlArg::F_ADD_SEALS(
+                SealFlag::F_SEAL_SEAL
+                    | SealFlag::F_SEAL_SHRINK
```
Comment (i509VCB) [inline src/wayland/seat/keyboard/keymap_file.rs:84]: Nix only appears to enable this on android and Linux.

Also we do need to test things compile on more than just Linux at some point lol

Comment (PolyMeilex) [reply src/wayland/seat/keyboard/keymap_file.rs:84]: [FreeBSD manual](https://www.freebsd.org/cgi/man.cgi?query=fcntl&sektion=2) list those, so I guess it's time to report a bug in `nix`

File: src/wayland/seat/keyboard/keymap_file.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            error!(log, "Error when creating sealed keymap file: {}", err);
+        }
+
+        Self {
+            sealed: sealed.ok(),
+            keymap,
+        }
+    }
+
+    pub fn with_fd<F>(&self, supports_seeled: bool, cb: F) -> Result<(), std::io::Error>
```
Comment (vberger) [inline src/wayland/seat/keyboard/keymap_file.rs:35]:
```suggestion
    pub fn with_fd<F>(&self, supports_sealed: bool, cb: F) -> Result<(), std::io::Error>
```

**Review by vberger — APPROVED** [review]
LGTM, just one small typo.

File: src/wayland/seat/keyboard/keymap_file.rs (hunk 3 above)
Comment (vberger) [reply src/wayland/seat/keyboard/keymap_file.rs]: Yeah, this needs some followup updates in the file too :sweat_smile:

Comment (PolyMeilex) [reply src/wayland/seat/keyboard/keymap_file.rs]: Done :sweat_smile:

**Pull request merged**